### PR TITLE
Add button to download static scoreboard ZIP. Fixes #1595.

### LINKED ISSUE #1595: Add way to download static scoreboard with all linked files
Add an export to export the static scoreboard together with:
* All CSS
* All Javascript
* All images
* A way to view team pages, either creating separate pages or doing something smart with some JSON. The latter is probably cleaner

Requested by @verwoerd and I like the idea.

### PR DESCRIPTION
**Pull request by nickygerritsen**
Team detail pages are not here (yet), since that is more difficult. If really wanted, we could add it, but this should already be a good start.

### CODE REVIEW

File: webapp/src/Controller/PublicController.php [hunk 1]
```diff
@@ -116,6 +95,114 @@ public function scoreboardAction(Request $request): Response
...
+                $contest = null;
+                foreach ($this->dj->getCurrentContests(-1) as $possibleContest) {
+                    if ($possibleContest->getCid() == $contestId || $possibleContest->getExternalid() == $contestId) {
+                        $contest = $possibleContest;
+                        break;
+                    }
+                }
+
+                if (!$contest) {
+                    throw new NotFoundHttpException('Specified contest not found');
```
Comment (vmcj) [inline webapp/src/Controller/PublicController.php:198]:
```suggestion
                    throw new NotFoundHttpException('Specified contest not found.');
```

You also add it for the error in https://github.com/DOMjudge/domjudge/pull/1641/files#diff-cc71aa6a9b851ee37894e1943a11fd1d6676ae347fac6e088ec93325e4a90e1aR130

Comment (nickygerritsen) [reply webapp/src/Controller/PublicController.php:198]: Well no, I copied it, but it makes sense.

File: webapp/src/Controller/PublicController.php [hunk 2]
```diff
@@ -116,6 +95,114 @@ public function scoreboardAction(Request $request): Response
...
+    {
+        $contest = null;
+        // For static scoreboards, allow to pass a contest= param.
+        if ($contestId = $request->query->get('contest')) {
+            if ($contestId === 'auto') {
+                // Automatically detect the contest that is activated the latest.
+                $contest      = null;
+                $activateTime = null;
+                foreach ($this->dj->getCurrentContests(-1) as $possibleContest) {
+                    if (!$possibleContest->getPublic() || !$possibleContest->getEnabled()) {
```
Comment (vmcj) [inline webapp/src/Controller/PublicController.php:179]:
```suggestion
                    if (!($possibleContest->getPublic() && $possibleContest->getEnabled())) {
```
I find this easier to read (I think you just copied this from what was there already)

File: webapp/src/Controller/PublicController.php [hunk 3]
```diff
@@ -116,6 +95,114 @@ public function scoreboardAction(Request $request): Response
...
+                        continue;
+                    }
+                    if ($activateTime === null || $activateTime < $possibleContest->getActivatetime()) {
+                        $activateTime = $possibleContest->getActivatetime();
+                        $contest      = $possibleContest;
+                    }
+                }
+            } else {
+                // Find the contest with the given ID.
+                $contest = null;
```
Comment (vmcj) [inline webapp/src/Controller/PublicController.php:189]: I think you already set this to null on L171

**Review by vmcj — APPROVED** [review]
Looks good to me, assuming you also tested it.

There are some variable initializations which can be moved outside of the if/else to save some lines.

File: webapp/src/Controller/PublicController.php [hunk 4]
```diff
@@ -116,6 +95,112 @@ public function scoreboardAction(Request $request): Response
         return $this->render('public/scoreboard.html.twig', $data, $response);
     }
 
+    /**
+     * @Route("/scoreboard-data.zip", name="public_scoreboard_data_zip")
+     */
+    public function scoreboardDataZipAction(RequestStack $requestStack, string $projectDir, Request $request): Response
+    {
+        $contest = $this->getContestFromRequest($request) ?? $this->dj->getCurrentContest(-1);
```
Comment (meisterT) [inline webapp/src/Controller/PublicController.php:103]: no access check required or is this implicit?

Comment (nickygerritsen) [reply webapp/src/Controller/PublicController.php:103]: The only check is for non public scoreboards and that should be implicit, but I will verify

Comment (nickygerritsen) [reply webapp/src/Controller/PublicController.php:103]: Indeed, `getCurrentContest` (which is also used in `getContestFromRequest`) checks whether you have access.

**Pull request merged**